### ISSUE TITLE
[Bug]: 7.1 Table of Contents active heading highlight broken

### ISSUE THREAD

**Issue by vmizgiris-equinix** [storybookjs/storybook#23584]
### Describe the bug

When there are subheadings (H3) with the same name under different headings (H2), the highlighting becomes broken, and only subheadings from the first H2 group get highlighted in the TOC as you scroll through the page.

### To Reproduce

1. Create a documentation page of a similar format:

```md
## Button

### Properties

Lorem ipsum

### Slots

Lorem ipsum

### Events

Lorem ipsum

## Input

### Properties

Lorem ipsum

### Slots

Lorem ipsum

### Events

Lorem ipsum

## Tooltip

### Properties

Lorem ipsum

### Slots

Lorem ipsum

### Events

Lorem ipsum
```

2. Add to a Storybook instance and enable TOC
3. Observe TOC links becoming active, only links from the first group "Button" will be highlighted as you scroll past "Input" and "Tooltip"

### System

```shell
System:
    OS: macOS 13.4.1
    CPU: (8) x64 Intel(R) Core(TM) i7-1068NG7 CPU @ 2.30GHz
  Binaries:
    Node: 16.18.1 - ~/Library/Caches/fnm_multishells/66230_1690247671085/bin/node
    npm: 9.7.1 - ~/Library/Caches/fnm_multishells/66230_1690247671085/bin/npm
  Browsers:
    Chrome: 115.0.5790.102
    Edge: 114.0.1823.67
    Safari: 16.5.2
  npmPackages:
    @storybook/addon-a11y: ^7.1.0 => 7.1.0 
    @storybook/addon-essentials: ^7.1.0 => 7.1.0 
    @storybook/addon-links: ^7.1.0 => 7.1.0 
    @storybook/addons: ^7.1.0 => 7.1.0
    @storybook/theming: ^7.1.0 => 7.1.0 
    @storybook/web-components: ^7.1.0 => 7.1.0 
    @storybook/web-components-webpack5: ^7.1.0 => 7.1.0
```

**Comment by bijumuhammad** [comment]
Hi,id like to help out as this would be a great starting point,though im facing issues while setting up this in local to reproduce the issue, im trying to troubleshoot that,Thank you

**Comment by Ihor-Lavrov** [comment]
> Hi,id like to help out as this would be a great starting point,though im facing issues while setting up this in local to reproduce the issue, im trying to troubleshoot that,Thank you


Hey, how are you doing? Any upd so far?

**Comment by bijumuhammad** [comment]
Yeah, sorry for the delay will give an update this weekend itself

**Comment by divyansh21jain** [comment]
@Ihor-Lavrov  Please correct me if I am wrong, what i understood is the problem is that only subheadings from the first H2 group are getting highlighted in the TOC as you scroll through the page, right?

To fix this issue, i guess we need to ensure that each subheading (H3) has a unique anchor link assigned to it, even if they have the same name under different H2 headings. One way to achieve this is by appending some unique identifier (e.g., a slug or numerical value) to the anchor link of each subheading.

Hope this might help. Happy to contribute!

**Comment by Sidnioulz** [comment]
@vmizgiris-equinix are you still able to reproduce this bug in Storybook 8.4?

**Comment by TheSeydiCharyyev** [comment]
Cannot reproduce on `next` (10.5.0-alpha).

Heading IDs are deduplicated on both rendering paths, so duplicate H3 names across different H2 sections get distinct anchors:

- Markdown headings in MDX go through `rehype-slug` (`code/addons/docs/src/mdx-plugin.ts`), which uses `github-slugger` and dedupes per document.
- `<Heading>` / `<Subheading>` blocks share a single `github-slugger` instance via `DocsSluggerContext`, created once per docs page in `DocsContainer` (`code/addons/docs/src/blocks/blocks/DocsContainer.tsx:29`). This landed in `baf7ac52717` (`fix(addons-docs): scope slugger to docs render`).

For the repro layout (`### Properties` / `### Slots` / `### Events` repeated under three H2s), a shared slugger produces `properties`, `properties-1`, `properties-2`, … — all 12 IDs distinct. Since tocbot's scrollspy keys off element IDs, distinct anchors are the prerequisite for per-section highlighting; the collision that caused the original bug no longer happens.

Safe to close unless someone can still reproduce on the current `next`.

**Comment by TheSeydiCharyyev** [comment]
@Sidnioulz — this answers your 8.4 repro question above: on `next` (10.5) the duplicate-H3 case no longer reproduces. Heading IDs are deduplicated on both render paths — markdown via `rehype-slug`, and `<Heading>`/`<Subheading>` blocks via the shared `github-slugger` in `DocsSluggerContext` — so same-named H3s under different H2s get distinct anchors and the TOC highlights correctly. Looks safe to close as fixed.
